pkg/apis/batch/validation/validation.go

```diff
@@ -486,9 +499,115 @@ func validatePodTemplateUpdate(spec, oldSpec batch.JobSpec, fldPath *field.Path,
+			}
+		}
+		if oldJob.Status.StartTime != nil && !ptr.Deref(job.Spec.Suspend, false) {
+			if job.Status.StartTime == nil {
+				allErrs = append(allErrs, field.Required(statusFld.Child("startTime"), "startTime cannot be removed for unsuspended job"))
+			}
+		}
+		if !ptr.Equal(oldJob.Status.StartTime, job.Status.StartTime) || !ptr.Equal(oldJob.Status.CompletionTime, job.Status.CompletionTime) {
+			if job.Status.StartTime != nil && job.Status.CompletionTime != nil && job.Status.CompletionTime.Before(job.Status.StartTime) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("completionTime"), job.Status.CompletionTime, "completionTime cannot be set before startTime"))
+			}
+		}
+		// validation when transitioning to terminal phase
+		for _, cType := range []batch.JobConditionType{batch.JobFailed, batch.JobComplete} {
+			if !isConditionTrue(oldJob.Status.Conditions, cType) && isConditionTrue(job.Status.Conditions, cType) {
+				if job.Status.StartTime == nil {
+					allErrs = append(allErrs, field.Required(statusFld.Child("startTime"), fmt.Sprintf("startTime is required for jobs with terminal %s=True condition", cType)))
+				}
+				if job.Status.Active > 0 {
+					allErrs = append(allErrs, field.Invalid(statusFld.Child("active"), job.Status.Active, fmt.Sprintf("active>0 is invalid for jobs with terminal %s=True condition", cType)))
+				}
+				if ptr.Deref(job.Status.Terminating, 0) > 0 {
+					allErrs = append(allErrs, field.Invalid(statusFld.Child("terminating"), *job.Status.Terminating, fmt.Sprintf("terminating>0 is invalid for jobs in terminal %s=True condition", cType)))
+				}
+				if job.Status.UncountedTerminatedPods != nil && len(job.Status.UncountedTerminatedPods.Failed)+len(job.Status.UncountedTerminatedPods.Succeeded) > 0 {
+					allErrs = append(allErrs, field.Invalid(statusFld.Child("uncountedTerminatedPods"), job.Status.UncountedTerminatedPods, fmt.Sprintf("uncountedTerminatedPods needs to be empty for jobs in terminal %s=True condition", cType)))
+				}
+			}
+		}
+		// validation to make sure Complete=True does not occur with Failure conditions
+		for _, failureType := range []batch.JobConditionType{batch.JobFailed, batch.JobFailureTarget} {
+			if isConditionTrue(job.Status.Conditions, batch.JobComplete) && isConditionTrue(job.Status.Conditions, failureType) {
+				if !isConditionTrue(oldJob.Status.Conditions, batch.JobComplete) || !isConditionTrue(oldJob.Status.Conditions, failureType) {
+					allErrs = append(allErrs, field.Invalid(statusFld.Child("conditions"), field.OmitValueType{}, fmt.Sprintf("both Complete=True and %s=True conditions cannot be set", string(failureType))))
+				}
+			}
+		}
+		if isIndexed && job.Spec.Completions != nil {
+			if job.Status.CompletedIndexes != "" {
+				if oldJob.Status.CompletedIndexes != job.Status.CompletedIndexes || oldJob.Status.Succeeded != job.Status.Succeeded {
```

Comment:
I'm not clear on why .status.Succeeded influences the check of completedIndexes.  If it's needed, add comment.  If it's not needed, remove.

Reply:
Oh, actually, this is a result of my experimentation, I clean up for now, the code wasn't complete.

Let me explain why I started this:  this is because the job controller for indexed jobs derives the Succeeded field from the count of `completedIndexed`, here: https://github.com/kubernetes/kubernetes/blob/d8122df1b73d0d92ebc32a7be9d8128d33f13dab/pkg/controller/job/job_controller.go#L1100-L1101. 

However, the analogous is not true for Failed indexes. Once I realized that I dropped the check in the middle (not finished). For failed indexes, it is possible that `|failedIndexes|<|failed|`, because it may take a couple of pods per index to classify an index as failed (when backoffLimitPerIndex is used). 

The opposite,  `|failedIndex|>|failed|`, may also happen when `backoffLimitPerIndex=0`, and we observe a failed pod. In that case we first report the failed pod in `uncountedTerminatedPods.Failed` (and move to `Failed` counter only after removing its finalizer. 

So I think what is true is `|failedIndex|<=|failed|+|uncountedTerminatedPods.Failed|`, but IMO it adds signifficant cognitive complexity, and just checking that the counter is non-decreasing should provide good validation. We also need to be careful not to make a bug by trying too validate too hard, as in case IMO.